File: Data Structures & Algorithms/trapping-rain-water/submission-3.java
class Solution {
    public int trap(int[] height) {
    int left = 0;
    int right = height.length-1; 

    int leftMax = height[0]; 
    int rightMax = height[height.length-1]; 
    int result = 0; 

    while(left < right){ 
        if(leftMax <= rightMax){ 
            left++; 
            leftMax = Math.max(leftMax, height[left]); 
            result += leftMax - height[left]; 
        } else { 
            right--; 
            rightMax = Math.max(rightMax, height[right]); 
            result += rightMax - height[right];
        }
    }


    return result;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if(not root): 
            return True
        
        def dfs(curr):

            if(not curr): 
                return 0 
            
            left = dfs(curr.left) + 1 
            right = dfs(curr.right) + 1

            height = max(left, right)

            return height

        leftHeight = dfs(root.left)
        rightHeight = dfs(root.right)

        if(abs(leftHeight - rightHeight) > 1): 
            return False

        return True